src/browser_action/on-each-interaction.js

```diff
@@ -11,43 +11,139 @@
  limitations under the License.
 */
 
+
+let recentLoAFs = [];
+
+const getIntersectingLoAFs = (start, end) => {
+  const intersectingLoAFs = [];
+
+  for (let i = 0, loaf; (loaf = recentLoAFs[i]); i++) {
+    // If the LoAF ends before the given start time, ignore it.
+    if (loaf.startTime + loaf.duration < start) continue;
+
+    // If the LoAF starts after the given end time, ignore it and all
+    // subsequent pending LoAFs (because they're in time order).
+    if (loaf.startTime > end) break;
+
+    // Still here? If so this LoAF intersects with the interaction.
+    intersectingLoAFs.push(loaf);
+  }
+  return intersectingLoAFs;
+};
+
+const loafObserver = new PerformanceObserver((list) => {
+
+  for (const entry of list.getEntries()) { recentLoAFs.push(entry) };
+  // We report interactions immediately, so don't need to keep many LoAFs around.
+  // Let's keep the last 5.
+  recentLoAFs = recentLoAFs.slice(-5);
+
+});
+loafObserver.observe({
+  type: 'long-animation-frame',
+  buffered: true,
+});
+
 /**
+ * We emulate an INP entry with similar logic to web-vitals.js
+ * But we have it easier as will emit it immediately and don't need the p98 stuff
+ * So can strip a lot of that logic out.
  * @param {Function} callback
  */
 export function onEachInteraction(callback) {
+
   const valueToRating = (score) => score <= 200 ? 'good' : score <= 500 ? 'needs-improvement' : 'poor';
 
   const observer = new PerformanceObserver((list) => {
-    const interactions = {};
 
-    for (const entry of list.getEntries().filter((entry) => entry.interactionId)) {
-      interactions[entry.interactionId] = interactions[entry.interactionId] || [];
-      interactions[entry.interactionId].push(entry);
+    const allEvents = list.getEntries();
+
+    // filter just for INP-eligible events (those with an interactionId)
+    const interactions = allEvents.filter((entry) => entry.interactionId)
+    // If none then can just end
+    if (interactions.length == 0 ) return;
+
+    const longestEntry = interactions.reduce((prev, curr) => prev.duration >= curr.duration ? prev : curr);
+    const largestDuration = longestEntry.duration;
+    const largestRenderTime = longestEntry.startTime + longestEntry.duration;
+
+    const longestFrameEntries = [];
+    const group = {
```

Comment:
Nit: the web-vitals library uses this "group" object to store a list of entries, but it doesn't look like you're doing that here, so you don't necessarily need to have these as properties of an object.

Reply:
So you're suggesting just to have three variables, rather than an object with three properties?

Reply:
Moved to that.